mp/src/game/shared/momentum/mom_gamemovement.cpp

```diff
@@ -328,6 +331,119 @@ void CMomentumGameMovement::WalkMove()
+    TryPlayerMove(&vecEndPos, &trace);
+
+    // The original StepMove will use the up result but with the down result's z-velocity if the up result goes farther.
+    // This generates z-velocity from nowhere, which means it is beneficial to hit uphill slopes without pressing jump
+    // to do ground movement instead of air movement. Instead, we should use one entire result or the other instead of
+    // combining them. The main reason we might want to use the down result over the up result even if the up result
+    // goes farther is if the down result causes the player to not be grounded in the future.
+    if (!sv_nojump_slope_boosts.GetBool() && mv->m_vecVelocity.z > NON_JUMP_VELOCITY)
+    {
+        float flStepDist = mv->GetAbsOrigin().z - vecPos.z;
+        if (flStepDist > 0.0f)
+        {
+            mv->m_outStepHeight += flStepDist;
+        }
+        return;
+    }
+
+    // Down results.
+    Vector vecDownPos, vecDownVel;
+    VectorCopy(mv->GetAbsOrigin(), vecDownPos);
+    VectorCopy(mv->m_vecVelocity, vecDownVel);
+
+    // Reset original values.
+    mv->SetAbsOrigin(vecPos);
+    VectorCopy(vecVel, mv->m_vecVelocity);
+
+    // Move up a stair height.
+    VectorCopy(mv->GetAbsOrigin(), vecEndPos);
+    if (player->m_Local.m_bAllowAutoMovement)
+    {
+        vecEndPos.z += player->m_Local.m_flStepSize + DIST_EPSILON;
+    }
+
+    TracePlayerBBox(mv->GetAbsOrigin(), vecEndPos, PlayerSolidMask(), COLLISION_GROUP_PLAYER_MOVEMENT, trace);
+    if (!trace.startsolid && !trace.allsolid)
+    {
+        mv->SetAbsOrigin(trace.endpos);
+    }
+
+    // Slide move up.
+    TryPlayerMove();
+
+    // Move down a stair (attempt to).
+    VectorCopy(mv->GetAbsOrigin(), vecEndPos);
+    if (player->m_Local.m_bAllowAutoMovement)
+    {
+        vecEndPos.z -= player->m_Local.m_flStepSize + DIST_EPSILON;
+    }
+
+    TracePlayerBBox(mv->GetAbsOrigin(), vecEndPos, PlayerSolidMask(), COLLISION_GROUP_PLAYER_MOVEMENT, trace);
+
+    // If we are not on the ground any more then use the original movement attempt.
+    if (trace.plane.normal[2] < 0.7)
+    {
+        mv->SetAbsOrigin(vecDownPos);
+        VectorCopy(vecDownVel, mv->m_vecVelocity);
+        float flStepDist = mv->GetAbsOrigin().z - vecPos.z;
+        if (flStepDist > 0.0f)
+        {
+            mv->m_outStepHeight += flStepDist;
+        }
+        return;
+    }
+
+    // If the trace ended up in empty space, copy the end over to the origin.
+    if (!trace.startsolid && !trace.allsolid)
+    {
+        mv->SetAbsOrigin(trace.endpos);
+    }
+
+    // Copy this origin to up.
+    Vector vecUpPos;
+    VectorCopy(mv->GetAbsOrigin(), vecUpPos);
+
+    // decide which one went farther
+    float flDownDist =
+        (vecDownPos.x - vecPos.x) * (vecDownPos.x - vecPos.x) + (vecDownPos.y - vecPos.y) * (vecDownPos.y - vecPos.y);
+    float flUpDist =
+        (vecUpPos.x - vecPos.x) * (vecUpPos.x - vecPos.x) + (vecUpPos.y - vecPos.y) * (vecUpPos.y - vecPos.y);
```

Comment:
```suggestion
    float flDownDist = vecDownPos.DistToSqr(vecPos);
    float flUpDist = vecUpPos.DistToSqr(vecPos);
```

Reply:
Should be 2D, so flDownDist = (vecDownPos - vecPos).Length2DSqr() and similar.

Reply:
Good catch! Surprised there isn't a `DistTo2D`...

Reply:
The bulk of that StepMove override is the original code from CGameMovement, I only changed a couple things where you can see the convar checked. It wouldn't hurt to change it to use that briefer distance calculation you suggested, but technically that method also subtracts the z components unnecessarily.